diff --git a/dbc_extract3/generate.bat b/dbc_extract3/generate.bat
index fed61dbd96..aead9ff3fc 100644
--- a/dbc_extract3/generate.bat
+++ b/dbc_extract3/generate.bat
@@ -28,7 +28,7 @@ goto usage
 :okay

 py -3 %RUNFILE% -p %GTINPATH% -b %BUILD%                      %PTR% -t scale  -o %OUTPATH%\sc_scale_data%PTREXT%.inc
-py -3 %RUNFILE% -p %INPATH%   -b %BUILD% --hotfix=%CACHEFILE% %PTR% -t output %BATCHFILE%
+py -3 %RUNFILE% -p %INPATH%   -b %BUILD% %PTR% -t output %BATCHFILE%

 echo Done!

diff --git a/engine/player/sc_player.cpp b/engine/player/sc_player.cpp
index d2e2aceb67..4a0d10272d 100644
--- a/engine/player/sc_player.cpp
+++ b/engine/player/sc_player.cpp
@@ -11707,7 +11707,12 @@ player_collected_data_t::action_sequence_data_t::action_sequence_data_t( const a
   {
     for ( cooldown_t* c : p->cooldown_list )
     {
-      if ( c->down() )
+      if ( c->charges > 1 && c->recharge_event )
+      {
+        cooldown_list.emplace_back( c, c->current_charge,
+            ( c->current_charge_remains() + ( c->charges - c->current_charge - 1 ) * cooldown_t::cooldown_duration( c ) ) );
+      }
+      else if ( c->down() )
       {
         cooldown_list.emplace_back( c, c->charges, c->remains() );
       }
diff --git a/engine/report/sc_report_html_player.cpp b/engine/report/sc_report_html_player.cpp
index 026a492304..4474a1b856 100644
--- a/engine/report/sc_report_html_player.cpp
+++ b/engine/report/sc_report_html_player.cpp
@@ -2318,6 +2318,52 @@ void print_html_sample_sequence_table_entry( report::sc_html_stream& os,
     }
   }

+  if ( p.specialization() == HUNTER_MARKSMANSHIP )
+  {
+    for ( const auto& cd_data : data.cooldown_list )
+    {
+      cooldown_t* cd = cd_data.object;
+      if ( cd->name_str == "aimed_shot" )
+      {
+        os.printf( ", %d (%.2f)", cd_data.value, cd_data.time_value.total_seconds() );
+      }
+    }
+  }
+
+  if ( p.specialization() == HUNTER_BEAST_MASTERY )
+  {
+    for ( const auto& cd_data : data.cooldown_list )
+    {
+      cooldown_t* cd = cd_data.object;
+      if ( cd->name_str == "barbed_shot" )
+      {
+        os.printf( ", %d (%.2f)", cd_data.value, cd_data.time_value.total_seconds() );
+      }
+    }
+  }
+
+  if ( p.specialization() == HUNTER_SURVIVAL )
+  {
+    for ( const auto& cd_data : data.cooldown_list )
+    {
+      cooldown_t* cd = cd_data.object;
+      if ( cd->name_str == "wildfire_bomb" )
+      {
+        os.printf( ", wfb %d (%.2f)", cd_data.value, cd_data.time_value.total_seconds() );
+      }
+
+      if ( cd->name_str == "butchery" )
+      {
+        os.printf( ", butch %d (%.2f)", cd_data.value, cd_data.time_value.total_seconds() );
+      }
+
+      if ( cd->name_str == "kill_command" )
+      {
+        os.printf( ", kc %d (%.2f)", cd_data.value, cd_data.time_value.total_seconds() );
+      }
+    }
+  }
+
   os << "</td>\n"
      << "<td class=\"left\">";
   first = true;
@@ -2325,6 +2371,7 @@ void print_html_sample_sequence_table_entry( report::sc_html_stream& os,
   {
     buff_t* buff = b_data.object;
     int stacks   = b_data.value;
+    double remains = b_data.time_value.total_seconds();

     if ( !buff->constant )
     {
@@ -2336,6 +2383,8 @@ void print_html_sample_sequence_table_entry( report::sc_html_stream& os,
       os.printf( "%s", util::encode_html( buff->name() ).c_str() );
       if ( stacks > 1 )
         os.printf( "(%d)", stacks );
+      if ( remains > 0 )
+        os.printf( "(%.2f)", remains );
     }
   }
   os << "</td>\n"
